src/util/kaldi-holder.cc

```diff
@@ -37,41 +39,104 @@ bool ExtractObjectRange(const Matrix<Real> &input, const std::string &range,
     return false;
   }
-  std::vector<int32> row_range, col_range;
+
   bool status = true;
+
   if (splits[0] != ":")
-    status = SplitStringToIntegers(splits[0], ":", false, &row_range);
+    status = SplitStringToIntegers(splits[0], ":", false, row_range);
+
   if (splits.size() == 2 && splits[1] != ":") {
-    status = status && SplitStringToIntegers(splits[1], ":", false, &col_range);
+    status = status && SplitStringToIntegers(splits[1], ":", false, col_range);
   }
-  if (row_range.size() == 0) {
-    row_range.push_back(0);
-    row_range.push_back(input.NumRows() - 1);
+  if (row_range->size() == 0) {
+    row_range->push_back(0);
+    row_range->push_back(rows - 1);
   }
-  if (col_range.size() == 0) {
-    col_range.push_back(0);
-    col_range.push_back(input.NumCols() - 1);
+  if (col_range->size() == 0) {
+    col_range->push_back(0);
+    col_range->push_back(cols - 1);
   }
 
   // Length tolerance of 3 -- 2 to account for edge effects when
   // frame-length is 25ms and frame-shift is 10ms, and 1 for rounding effects
   // since segments are usually retained up to 2 decimal places.
   int32 length_tolerance = 3;
-  if (!(status && row_range.size() == 2 && col_range.size() == 2 &&
-        row_range[0] >= 0 && row_range[0] <= row_range[1] &&
-        row_range[1] < input.NumRows() + length_tolerance &&
-        col_range[0] >=0 &&
-        col_range[0] <= col_range[1] && col_range[1] < input.NumCols())) {
+  if (!(status && row_range->size() == 2 && col_range->size() == 2 &&
+        row_range->at(0) >= 0 && row_range->at(0) <= row_range->at(1) &&
+        row_range->at(1) < rows + length_tolerance &&
+        col_range->at(0) >=0 &&
+        col_range->at(0) <= col_range->at(1) && col_range->at(1) < cols)) {
     KALDI_ERR << "Invalid range specifier: " << range
-              << " for matrix of size " << input.NumRows()
-              << "x" << input.NumCols();
+              << " for matrix of size " << rows
+              << "x" << cols;
     return false;
   }
 
-  if (row_range[1] >= input.NumRows())
-    KALDI_WARN << "Row range " << row_range[0] << ":" << row_range[1]
+  if (row_range->at(1) >= rows)
+    KALDI_WARN << "Row range " << row_range->at(0) << ":" << row_range->at(1)
                << " goes beyond the number of rows of the "
-               << "matrix " << input.NumRows();
+               << "matrix " << rows;
+  return status;
+}
+
+bool ExtractObjectRange(const GeneralMatrix &input, const std::string &range,
+                        GeneralMatrix *output) {
+
+  Matrix<BaseFloat> out;
+  if (input.Type() == kFullMatrix) {
+    const Matrix<BaseFloat> &in = input.GetFullMatrix();
+    ExtractObjectRange(in, range, &out);
+  } else if (input.Type() == kCompressedMatrix) {
+    const CompressedMatrix &in = input.GetCompressedMatrix();
+    ExtractObjectRange(in, range, &out);
+  } else if (input.Type() == kSparseMatrix) {
+    Matrix<BaseFloat> in, out;
```

Comment:
please put a comment here mentioning that this is very ineffcient,  and if it ever becomes a bottleneck we'll implement something smarter.
Also, 'out' doesn't need to be redeclared.  And I don't like the variables 'in' and 'input' existing, since the two words are synonyms.  I'd prefer in_mat vs in.